Support running the same application multiple times in dev mode

Postremus opened the pull request:
Adds support for running the same application multiple times in dev mode on the same machine.
Problem was, that resources of webjars (like devui) where unpacked to %TEMP%/quarkus/applicationName/libraryName/libraryVersion.
On windows, this worked fine. %TEMP% points to a path inside the user directory (e.g. c:/users/martin/appdata/local/temp/).
On linux however, this caused AccessDeniedException, since %TEMP% points to a global temp directory, e.g. /temp.

This branch changes the location where webjars are unpacked to.
Now, every quarkus dev mode isntace creates a new temp directory, e.g. C:\Users\Martin\AppData\Local\Temp\quarkus-webjar10464518099488497426.
These temp directories are removed on shutdown.


The directory layout inside the temp directory is changed. It is now flatter, and the same as in the generated resource. e.g. quarkus-webjar10464518099488497426\META-INF\swagger-ui-files\index.html

I also deprecated io.quarkus.deployment.util.WebJarUtil (forRemoval) in favor of two new BuildItems - WebJarBuildItem and WebJarResultBuildItem.
WebJarBuildItem contains the instructions on which artifact should be unpacked and to which final destination (e.g. META-INF/swagger-ui-files). The webjar is either unpacked to disk (dev mode), or resources are generated (prod).
WebJarResultBuildItem contains the path where the webjar was unpacked to.

dev-ui, swagger-ui, smallrye-health and graphql already use these new builditems - in quarkus, no place uses the deprecated WebJarUtil anymore.

fixes #22515

Related zulip discussion: https://quarkusio.zulipchat.com/#narrow/stream/187038-dev/topic/Refactoring.20WebJarUtil

Postremus commented on extensions/swagger-ui/deployment/src/main/java/io/quarkus/swaggerui/deployment/SwaggerUiProcessor.java:
```diff
@@ -181,14 +158,23 @@ public void getSwaggerUiFinalDestination(
-            Handler<RoutingContext> handler = recorder.handler(finalDestinationBuildItem.getSwaggerUiFinalDestination(),
-                    finalDestinationBuildItem.getSwaggerUiPath(),
+            String swaggerUiPath = nonApplicationRootPathBuildItem.resolvePath(swaggerUiConfig.path);
+            swaggerUiBuildProducer.produce(new SwaggerUiBuildItem(result.getFinalDestination(), swaggerUiPath));
```
`SwaggerUiBuildItem` just to be on the safe side, no idea if anyone consumes it.

Maybe we can deprecate it also?

Postremus commented on extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java:
```diff
@@ -478,14 +445,24 @@ void registerHealthUiHandler(
-                    smallRyeHealthBuildItem.getHealthUiPath(), runtimeConfig);
+            String healthUiPath = nonApplicationRootPathBuildItem.resolvePath(healthConfig.ui.rootPath);
+            smallryeHealthBuildProducer
+                    .produce(new SmallRyeHealthBuildItem(result.getFinalDestination(), healthUiPath));
```
`SmallRyeHealthBuildItem` just to be on the safe side, no idea if anyone consumes it.

Maybe we can deprecate it also?

Postremus commented:
This PR is ready for review - it contains everything I wanted to include for now.

Dev mode startup time worsens by about 30ms. According to profiling, this is because the temp directory is now deleted on shutdown, and webjar resources need to be copied again.

I plan to open a follow up PR once this one is merged which will implement #22525 for swagger-ui, smallrye-health, graphql.
Should be enough to counteract these 30ms.

Postremus commented on extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallRyeGraphQLProcessor.java:
```diff
@@ -603,14 +567,24 @@ void registerGraphQLUiHandler(
-            Handler<RoutingContext> handler = recorder.uiHandler(smallRyeGraphQLBuildItem.getGraphqlUiFinalDestination(),
-                    smallRyeGraphQLBuildItem.getGraphqlUiPath(), runtimeConfig);
+            String graphQLUiPath = nonApplicationRootPathBuildItem.resolvePath(graphQLConfig.ui.rootPath);
+            smallRyeGraphQLBuildProducer
```
`SmallRyeGraphQLBuildItem` is produced just to be on the safe side, no idea if anyone consumes it.

Maybe we can deprecate the build item as well?

aloubyansky commented on extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/webjar/WebJarUtil.java:
```diff
@@ -0,0 +1,273 @@
+
+        ClassLoader classLoader = WebJarUtil.class.getClassLoader();
+
+        for (Path p : resourcesArtifact.getResolvedPaths()) {
```
We should be moving away from using resolved path to explore the content of a resolved dependency switching to `PathTree userApplication.getContentTree()`. Then you walk the tree or process a specific path.

Postremus replied:
done

aloubyansky commented on extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/webjar/WebJarUtil.java:
```diff
@@ -0,0 +1,273 @@
+            String artifactId) {
+        for (ResolvedDependency dep : curateOutcomeBuildItem.getApplicationModel().getDependencies()) {
+            if (dep.getArtifactId().equals(artifactId)
+                    && dep.getGroupId().equals(groupId)) {
```
Keep in mind, it may happen there could be multiple artifacts on the classpath with the same G:A but different classifiers.

Postremus replied:
We now also compare the classifier in this place.
I saw no other way to compare a GACT to the GACTV of the ResolvedDependency.

aloubyansky commented on extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/webjar/WebJarUtil.java:
```diff
@@ -0,0 +1,285 @@
+
+        resourcesArtifact.getContentTree().accept(webJar.getRoot(), new Consumer<PathVisit>() {
+            @Override
+            public void accept(PathVisit pathVisit) {
```
Theoretically, the passed in `pathVisit` could be null if the content tree does not contain the path (webJar.getRoot), not sure if that could happen in this specific case though. I thought it would be up to the impl of the consumer whether to throw an error or simply return in such a case.

Postremus replied:
done

aloubyansky commented on extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/webjar/WebJarUtil.java:
```diff
@@ -0,0 +1,285 @@
+        for (ResolvedDependency dep : curateOutcomeBuildItem.getApplicationModel().getDependencies()) {
+            if (dep.getArtifactId().equals(artifactKey.getArtifactId())
+                    && dep.getGroupId().equals(artifactKey.getGroupId())
+                    && dep.getClassifier().equals(artifactKey.getClassifier())) {
```
That or `dep.getKey().equals(artifactKey)`

Postremus replied:
I thought about it, too, but I don't see how `.equals` can return true.
`ResolvedDependency.getKey()` returns an instance of GACTV.

We only pass a GACT in here, since the version of the webjar is not static.
E.g. io.quarkus: quarkus-vertx-http-deployment is one of it.

aloubyansky replied:
> I thought about it, too, but I don't see how `.equals` can return true. `ResolvedDependency.getKey()` returns an instance of GACTV.

No, it returns an `ArtifactKey`, which is `GACT`.

Postremus replied:
yeah, you are obviously right. I guess I got confused, because ResolvedArtifactDependency extends GACTV. :D

Postremus commented:
I cleaned this up a bit further, and implemented following additonal changes / fixes:
* `onlyCopyNonArtifactFiles` now has no effect in prod builds, since all files are needed anyway.
* insertVariables is now also a simple `WebJarResourcesFilter`
* `CombinedWebJarResoucesFilter` allows to combine multiple filters into one. This allows to propagate the `changed` flag through all filters. For example, SwaggerUIProcessor generated a changed index.html file. With `onlyCopyNonArtifactFiles=true`, this file was not copied to disk however, since no change was detected.

gsmet commented on extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/webjar/WebJarResultBuildItem.java:
```diff
@@ -0,0 +1,46 @@
+import io.quarkus.maven.dependency.ArtifactKey;
+import io.quarkus.maven.dependency.ResolvedDependency;
+
+public final class WebJarResultBuildItem extends SimpleBuildItem {
```
Could we have some class-level javadoc here?

gsmet replied:
I think I would call it `WebJarResultsBuildItem` plural. You have several and it would be less confusing.

gsmet commented on extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/devmode/console/DevConsoleProcessor.java:
```diff
@@ -123,7 +123,10 @@
+    private static final GACT DEVCONSOLE_WEBJAR_ARTIFACT_KEY = new GACT("io.quarkus", "quarkus-vertx-http-deployment", null,
+            "jar");
+    private static final String DEVCONSOLE_WEBJAR_STATIC_RESOURCES_PATH = "dev-static/";
+    private static final String DEVCONSOLE_FINAL_DESTINATION = "META-INF/dev-ui-files";
```
Why do we need this now? Couldn't it be generated from the artifact id/version + resource path? That would simplify things a bit.
At least I don't think we should make it mandatory.

Postremus replied:
I removed this option completly. Automaticly generating the final destination also prevents clashes between different webjars, which is quite nice.

gsmet commented on extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/webjar/WebJarBuildItem.java:
```diff
@@ -0,0 +1,76 @@
+        this(artifactKey, webjarRoot, true, false, finalDestination, filter);
+    }
+
+    public WebJarBuildItem(GACT webJarKey, String webjarRoot, boolean useDefaultQuarkusBranding,
```
I think it would be handy to have versions of both constructors without the filter as it's purely optional.

gsmet replied:
That being said, I wonder if we should use a Builder approach for this one?

gsmet replied:
Maybe rename `webJarKey` to `artifactKey` for consistency?

gsmet commented on core/deployment/src/main/java/io/quarkus/deployment/util/WebJarUtil.java:
```diff
@@ -44,7 +44,10 @@
 /**
  * Utility for Web resource related operations
+ * 
+ * @deprecated Use WebJarBuildItem and WebJarResultBuildItem instead.
```
If you rename `WebJarResultBuildItem`, make sure this is done everywhere including here.

gsmet commented on extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/webjar/WebJarBuildItem.java:
```diff
@@ -0,0 +1,76 @@
+    private final boolean onlyCopyNonArtifactFiles;
+
+    /**
+     * Defines whether quarkus can override resources of the webjar with quarkus internal files.
```
```suggestion
     * Defines whether Quarkus can override resources of the webjar with Quarkus internal files.
```

gsmet reviewed (changes requested):
Finally had the time to look at this one. I also migrated the Quarkus GitHub App to it to make sure things would be smooth.

Things are indeed far better with this setup and I was able to remove quite some boilerplate that was very ad hoc.

So it's a big +1 from me but... I added a few suggestions. They are open to discussion so feel free to push back :).

Postremus commented:
Thank you for the review @gsmet.
I implemented every change.

Postremus commented:
test failures look related, will take a closer look soon.

gsmet commented:
@Postremus just a note that the cutoff for 2.7.0.CR1 is tonight and I would prefer having this one in CR1 given how structural it is.

Postremus committed:
Support running the same application multiple times in dev mode

Postremus commented:
Should be fixed now, lets see what CI has to say.

gsmet commented:
And all green! Thanks for your efforts on this @Postremus, it's much cleaner!
